iothub/device/samples/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs

```diff
@@ -147,93 +283,119 @@ private Task<MethodResponse> HandleMaxMinReportCommand(MethodRequest request, ob
-                            $"endTime={report.endTime.LocalDateTime}");
+                            $" maxTemp={report.MaximumTemperature}°C, minTemp={report.MinimumTemperature}°C, avgTemp={report.AverageTemperature}°C," +
+                            $" startTime={report.StartTime.LocalDateTime}, endTime={report.EndTime.LocalDateTime}");
 
-                        byte[] responsePayload = Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(report));
-                        return Task.FromResult(new MethodResponse(responsePayload, (int)StatusCode.Completed));
+                        return Task.FromResult(new CommandResponse(report, (int)StatusCode.Completed, s_payloadConvention));
                     }
 
                     _logger.LogDebug($"Command: component=\"{componentName}\", no relevant readings found since {sinceInDateTimeOffset.LocalDateTime}, " +
                         $"cannot generate any report.");
-                    return Task.FromResult(new MethodResponse((int)StatusCode.NotFound));
+                    return Task.FromResult(new CommandResponse((int)StatusCode.NotFound));
                 }
 
                 _logger.LogDebug($"Command: component=\"{componentName}\", no temperature readings sent yet, cannot generate any report.");
-                return Task.FromResult(new MethodResponse((int)StatusCode.NotFound));
+                return Task.FromResult(new CommandResponse((int)StatusCode.NotFound));
             }
             catch (JsonReaderException ex)
             {
                 _logger.LogDebug($"Command input is invalid: {ex.Message}.");
-                return Task.FromResult(new MethodResponse((int)StatusCode.BadRequest));
+                return Task.FromResult(new CommandResponse((int)StatusCode.BadRequest));
             }
         }
 
-        private Task SetDesiredPropertyUpdateCallback(TwinCollection desiredProperties, object userContext)
+        private Task WritablePropertyEventDispatcherAsync(PropertyCollection desiredProperties, object cancellationToken)
         {
-            bool callbackNotInvoked = true;
-
             foreach (KeyValuePair<string, object> propertyUpdate in desiredProperties)
             {
-                string componentName = propertyUpdate.Key;
-                if (_desiredPropertyUpdateCallbacks.ContainsKey(componentName))
+                // The dispatcher key will be either the root-level property name or the component name.
+                string dispatcherKey = propertyUpdate.Key;
+                if (_writablePropertyEventCallbacks.ContainsKey(dispatcherKey))
                 {
-                    _desiredPropertyUpdateCallbacks[componentName]?.Invoke(desiredProperties, componentName);
-                    callbackNotInvoked = false;
+                    return _writablePropertyEventCallbacks[dispatcherKey]?.Invoke(desiredProperties, cancellationToken, dispatcherKey);
```

Comment:
don't need to send the entire property collection, just send "KeyValuePair<string, object> propertyUpdate" and "PropertyCollection.Version".

Reply:
Done